### PR TITLE
Hog/experiment initial

### PR DESCRIPTION
**Pull request by ericlee0803**
- Add experiment api 
- Add associated integration test, which generates random experiment config 
- Fix naming convention in run API for "Test Set" and "Training Set" to be "TestSet" and "TrainingSet" to fix XGB error

Note that this PR, like the last one, is for feedback first. The integration test in particular needs to be beefed up but I would like suggestions on what else to test.

### CODE REVIEW

**Review by yuhong-sigopt — APPROVED** [review]

**Comment by ericlee0803** [comment]
Also worth noting; it's probably easier to beef up the integration tests in a separate PR like last time

File: sigopt/xgboost/experiment.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    experiment_config_parsed = copy.deepcopy(self.experiment_config)
+
+    # Check experiment config optimization metric
+    for metric in experiment_config_parsed['metrics']:
+      if metric['strategy'] == 'optimize':
+        assert metric['name'] in DEFAULT_CLASSIFICATION_METRICS or metric['name'] in DEFAULT_REGRESSION_METRICS
+
+        # change optimized metric to reflect updated name
+        metric['name'] = self.evals[0][1] + '-' + metric['name'] if isinstance(self.evals, list) else \
+          DEFAULT_EVALS_NAME + '-' + metric['name']
```
Comment (harvey-intel) [inline sigopt/xgboost/experiment.py:36]: syntax, we don't do that `\` for line break. Just change this to a regular if block
```
if isinstance(...):
  metric['name'].......
else:
  ...
```

File: sigopt/xgboost/experiment.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    # Check key overlap between parameters to be optimized and parameters that are set
+    params_optimized = [param['name'] for param in self.experiment_config['parameters']]
+    assert len(set(params_optimized) & set(self.params.keys())) == 0, \
```
Comment (harvey-intel) [inline sigopt/xgboost/experiment.py:40]: against no `\` for line break, this can be changed with parenthesis
```
assert ......., (
  'There is ....'
)
  ```

File: sigopt/xgboost/experiment.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      'There is overlap between optimized params and user-set params'
```
Comment (harvey-intel) [inline sigopt/xgboost/experiment.py:41]: `There is overlap between tuned parameters (through `experiment_config`) and user-set parameters (through `params`). Parameter name cannot be defined in both locations.` 

Ideally, I want this message to be even more informative, like listing out the parameter names that are overlapping. Don't have to have it right now. But can you add it as a TODO comment  here?

File: sigopt/xgboost/experiment.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  def run_experiment(self):
+    for run in self.sigopt_experiment.loop():
+      with run:
+        if self.num_boost_round:
+          num_boost_round_run = self.num_boost_round
+        elif 'num_boost_round' in run.params:
+          num_boost_round_run = run.params['num_boost_round']
+        else:
+          num_boost_round_run = DEFAULT_NUM_BOOST_ROUND
```
Comment (harvey-intel) [inline sigopt/xgboost/experiment.py:57]: seems a bit messy here, but ok.

File: sigopt/xgboost/experiment.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+          if name not in PARAMS_LOGGED_AS_METADATA:
+            run.params.update({name: self.params[name]})
+        if 'num_boost_round' not in run.params:
+          run.params.update({'num_boost_round': num_boost_round_run})
+
+
+        XGBRun(
+          all_params, self.dtrain, num_boost_round=num_boost_round_run,
+          evals=self.evals, run_options=self.run_options
+        )
```
Comment (harvey-intel) [inline sigopt/xgboost/experiment.py:73]: syntax: one parameter per line.

File: integration_test/test_xgboost_experiment.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
+import random
+import pytest
+
+import sigopt.xgboost
+from .test_xgboost import (
+  _form_random_run_params
+)
```
Comment (harvey-intel) [inline integration_test/test_xgboost_experiment.py:7]: alphabetize your imports

File: integration_test/test_xgboost_experiment.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+SEARCH_SPACES = [
+  {
+      'name': 'eta',
+      'type': 'double',
+      'bounds': {'min': 0.1, 'max': 0.5}
```
Comment (harvey-intel) [inline integration_test/test_xgboost_experiment.py:13]: indentation.

File: integration_test/test_xgboost_experiment.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    'bounds': {'min': 2, 'max': 5}
+  },
+  {
+    'name': 'num_boost_round',
+    'type': 'int',
+    'bounds': {'min': 2, 'max': 5}
+  },
+]
+
+from sigopt.xgboost.experiment import DEFAULT_CLASSIFICATION_METRICS, DEFAULT_REGRESSION_METRICS
```
Comment (harvey-intel) [inline integration_test/test_xgboost_experiment.py:32]: but this with the other imports.

File: integration_test/test_xgboost_experiment.py (hunk 6 above)
Comment (harvey-intel) [reply integration_test/test_xgboost_experiment.py:7]: This can just be a single import no need the 
```
(
  _form_...
)
```

**Review by harvey-intel — COMMENTED** [review]
Mostly looks good to me. Some nit pickings.

File: sigopt/xgboost/experiment.py (hunk 4 above)
Comment (ericlee0803) [reply sigopt/xgboost/experiment.py:57]: Dealing with the num_boost_round is a headache not sure if there's a better way of doing it

**Review by harvey-intel — APPROVED** [review]
thanks.

File: sigopt/xgboost/experiment.py (hunk 4 above)
Comment (yuhong-sigopt) [reply sigopt/xgboost/experiment.py:59]: I am not sure why `num_boost_round` is different from other parameters such as `eta`?  In my opinion, in `sigopt.xgboost.run.make_run`, we should merge parameters from `run.params` and user defined, so we can remove check parameters here.

File: sigopt/xgboost/experiment.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+      assert 'num_boost_round' not in params_optimized
+
+    self.sigopt_experiment = create_experiment(**experiment_config_parsed)
+
+  def run_experiment(self):
+    for run in self.sigopt_experiment.loop():
+      with run:
+        if self.num_boost_round:
+          num_boost_round_run = self.num_boost_round
+        elif 'num_boost_round' in run.params:
```
Comment (yuhong-sigopt) [inline sigopt/xgboost/experiment.py:56]: Why user define priority is higher than `run.params`?
